Improve load balancing scalability in swarm mode

ctelfer opened the pull request:
Docker swarm mode supports load balancing traffic to endpoints in a server using a distributed load balancing approach.  In this model, all containers a swarm network can act as load balancers for any service resident on a network that the container is present on.  This has the advantage of not requiring service traffic to cross the network twice (once to the load balancer and once to the endpoint) as in a more traditional architecture.  It also avoids dangers of having a single point of failure and the problems of having to scale the capacity of the load balancer itself.  Rather, the load balancing performance can scale as the compute power and network capacity of the scales.

The main problem with the current approach is its own strength.  It relies on distributing the load balancing configuration to every container in the swarm.  This means that the amount of state that each node must maintain grows as the square of the number of services in the swarm and the square of the average number of containers per service.

This patch conceptually borrows from the architecture of the Windows swarm service support by creating a load-balancer endpoint for each network on each node in the swarm.  Libnetwork assigns this endpoint its own IP address on the network as well as the VIPs of every service with on the network.  Every container on the node will send its service requests via the VIPs of the service which will therefore arrive at the load balancing endpoint.   This endpoint will perform load balancing using IPVS as was done in the service mesh except that the outgoing service request will use the endpoint's private IP address as the source address for the request (via NAT).  This is so that return traffic will first go to the load balancing endpoint in order to hit the load balancing state mapping and from there the Linux stack will redirect the response traffic back to the connection initiator.

While this approach drastically improves the load balancing scalability it has two potential negative impacts.   First, the original behavior included the ability to load balance to a service by sending to *any* task IP address for the service.  While convenient, this is not likely to be a widely used feature since the DNS entries for a service only point to the VIP and not to the backends.  Second, this scheme requires allocating extra IP addresses in each overlay network (one per node that has a task in the network).  If a network was near IP address exhaustion, this could cause an existing swarm network to break during an upgrade of the moby engine.  The problem can safely be avoided for all new networks by simply allocating sufficiently large address-blocks for each network. 

This patch is staged with a series smaller patches that preserve libnetwork's current functioning, followed by a larger patch that contains the main change itself.  The patch also includes the addition of a DeleteIngress() method to the Network interface to move load balancing cleanup logic from moby back into libnetwork where it belongs.   This was discussed in https://github.com/moby/moby/pull/36538.

Incorporating this into moby requires a small set of changes to enable the functionality in moby (and remove the deprecated functionality mentioned above).

selansen commented:
Looks like CI failed

ctelfer commented:
Sorry, forgot to run golint.  Adjusted for golint errors and re-pushed.

selansen commented on network.go:
```diff
@@ -957,10 +989,36 @@ func (n *network) delete(force bool) error {
+	if n.hasLoadBalancerEndpoint() {
+		emptyCount = 1
+	}
+	if !force && n.getEpCnt().EndpointCnt() > emptyCount {
```
Not sure why do you have emptyCount variable. Anyways you are using only one place. instead , you can directly use n.hasLoadBalancerEndpoint().

ctelfer replied:
The code is attempting to determine whether there are any *container& endpoints left in the network.  If there are, then it can not delete the network.  The code does this by querying `n.getEpCnt().EndpointCnt()` to get the current number of endpoints connected to the network.  If the network has a load balancer endpoint, then a count of 1 indicates that there are no container endpoints in the network.  If there is no load balancer endpoint, then only a count of 0 indicates that there are no more container endpoints.

selansen commented:
one genreal comment :  There are many new functions and changes , But I dont see any unit test. Is it possible to add unit test?

ctelfer commented on network.go:
```diff
@@ -957,10 +989,36 @@ func (n *network) delete(force bool) error {
+				return err
+			}
+			// continue deletion when force is true even on error
+			logrus.Warnf("Error deleting load balancer sandbox: %v", err)
```
I believe I was just keeping it consistent with the previous behavior which put it at the warning level.  Presumably this behavior was because failure to remove may be due to the fact that it already got inadvertantly removed.   In this case, the error is not fatal to the system since the network is being removed anyway.  (see the comment above the warning).  If this is the rationale, then warning level seems appropriate.  The error message was, I think, the same, but we could change the wording in principle.

ctelfer commented on service_linux.go:
```diff
@@ -112,9 +116,149 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
+		if sb.ingress {
+			var gwIP net.IP
+			if ep := sb.getGatewayEndpoint(); ep != nil {
+				gwIP = ep.Iface().Address().IP
```
Ah, except for the fact that the gwIP address needs to live outside of the scope of the if statement and should be its default value if  `sb.getGatewayEndpoint()` returns `nil`.  It's something akin to writing:
```go
var x int = 0
if something {
    x = 1
}
// Do something with 'x'
```
Except that golint complains if you initialize a newly declared variable to the default value.  In C it would look much nicer of course.  :)
```C
int x = something ? 1 : 0;
```

ctelfer commented:
Perhaps there is a way to do unit tests for a feature like this, but I'm not clear as to how myself.  To really test it we need containers, namespaces, etc..   (i.e. what moby does)

We currently don't have any unit tests that validate addServiceBinding().  We could create tests that would validate things like the service binding data structures themselves by stubbing out the backend code and that could be worthwhile.  But that won't validate *this* patchset because all the changes for this set are in the backend code.  But if you have some ideas about how we could approach this, please do suggest it.  Otherwise it will likely have to be done in integration tests w/ Moby.

fcrisciani commented on service_linux.go:
```diff
@@ -112,12 +116,149 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
+	if !ok {
+		return nil, nil, fmt.Errorf("Unable to get sandbox for %s(%s) in for %s", ep.Name(), ep.ID(), n.ID())
+	}
+	// XXX really?  really?  ugh
```
maybe we should remove this :D

ctelfer replied:
Rats.  Forgot to get rid of that!

ctelfer commented on network.go:
```diff
@@ -2055,8 +2055,20 @@ func (c *controller) getConfigNetwork(name string) (*network, error) {
+func (n *network) lbSandboxName() string {
+	name := n.name + "-sbox"
+	if n.ingress {
+		name = "lb-" + n.name
```
Doh!  That's a logic flip.  ingress was supposed to remain the same, while new endpoint sandboxes were supposed to change.   Will fix.

ctelfer commented:
Per review w/ @fcrisciani, updating this to remove the backward-compatible code.

selansen commented:
LGTM

fcrisciani commented on network.go:
```diff
@@ -2055,8 +2098,21 @@ func (c *controller) getConfigNetwork(name string) (*network, error) {
+}
+
+func (n *network) lbEndpointName() string {
+	return n.name + "-endpoint"
```
for the default gw we put a restriction on the max string length, do we need something similar for the lb endpoints or sandbox names?

ctelfer replied:
For some reason we restrict the length of the names for gateway *sandboxes* to 12 characters plus the prefix "gateway_".  I'm not sure why we do this for sandboxes only but there doesn't seem to be such a restriction for endpoints, so this should be safe.

fcrisciani commented on network.go:
```diff
@@ -2073,11 +2129,14 @@ func (n *network) createLoadBalancerSandbox() error {
 		CreateOptionLoadBalancer(),
 	}
+	if n.hasLoadBalancerEndpoint() && !n.ingress {
+		epOptions = append(epOptions, CreateOptionAnonymous())
```
I guess worth spending a comment saying that Anonymous is to avoid it to be resolvable from the DNS

fcrisciani commented on service_common.go:
```diff
@@ -291,9 +303,7 @@ func (c *controller) addServiceBinding(svcName, svcID, nID, eID, containerName s
 
 	// Add loadbalancer service and backend in all sandboxes in
```
maybe this comment now is outdated. It should be in the load balancer sandbox only, not all sandboxes

fcrisciani commented on service_common.go:
```diff
@@ -368,8 +390,8 @@ func (c *controller) rmServiceBinding(svcName, svcID, nID, eID, containerName st
 
 	// Remove loadbalancer service(if needed) and backend in all
 	// sandboxes in the network only if the vip is valid.
```
same here

fcrisciani commented on service_linux.go:
```diff
@@ -32,7 +32,7 @@ func init() {
 // Get all loadbalancers on this network that is currently discovered
 // on this node.
-func (n *network) connectedLoadbalancers() []*loadBalancer {
+func (n *network) connectedLoadBalancers() []*loadBalancer {
```
confused on the following 2 methods, are they still making sense? there should be only 1 LB per network right?

fcrisciani commented on service_linux.go:
```diff
@@ -77,96 +75,74 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
+	if !ok {
+		return nil, nil, fmt.Errorf("Unable to get sandbox for %s(%s) in for %s", ep.Name(), ep.ID(), n.ID())
+	}
+	ep = sb.getEndpoint(ep.ID())
```
aaarghh, every time that we need a logic like this, one part of me dies....

fcrisciani commented on service_linux.go:
```diff
@@ -77,96 +75,74 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
-				eIP = ep.Iface().Address()
+		if epi.LoadBalancer() {
+			var err error
+			ep, err = n.getEndpointFromStore(e.ID())
```
e is already coming from store, can we use that already?

fcrisciani commented on service_linux.go:
```diff
@@ -77,96 +75,74 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
+	return ""
 }
 
 // Add loadbalancer backend to all sandboxes which has a connection to
```
update comment, not `all` sandboxes

fcrisciani commented on service_linux.go:
```diff
@@ -77,96 +75,74 @@ func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
-
-// Add loadbalancer backend into one connected sandbox.
-func (sb *sandbox) addLBBackend(ip, vip net.IP, fwMark uint32, ingressPorts []*PortConfig, eIP *net.IPNet, gwIP net.IP, isIngressNetwork bool) {
+func (n *network) addLBBackend(ip net.IP, lb *loadBalancer) {
```
this is a not for myself, maybe I miss it but if the wiring of this method is the same as before, are we calling it for each container? if so the findLBEndpointSandbox is actually pretty expensive in term of loops that is doing for the same disconnection due to the existence of the store

ctelfer replied:
The merge made some of the changes here a bit confusing because of similarly named methods for both sandbox and network structures.  This function does indeed only add a single LB entry per network.  (Once the final patch is applied:  if one is only looking at the commit that refactors the API then one will still see the "walk endpoints" loop as that commit doesn't change the loadbalancing yet.)

fcrisciani commented on service_linux.go:
```diff
@@ -212,20 +203,32 @@ func (sb *sandbox) addLBBackend(ip, vip net.IP, fwMark uint32, ingressPorts []*P
+// network. If 'rmService' is true, then remove the service entry as well.
+// If 'fullRemove' is true then completely remove the entry, otherwise
+// just deweight it for now.
+func (n *network) rmLBBackend(ip net.IP, lb *loadBalancer, rmService bool, fullRemove bool) {
```
same for this method, are we calling it for each container?

ctelfer commented on network.go:
```diff
@@ -957,10 +975,36 @@ func (n *network) delete(force bool) error {
+	}
+
+	// Check that the network is empty
+	var emptyCount uint64 = 0
```
yep, will fix.

ctelfer commented on network.go:
```diff
@@ -42,6 +42,12 @@ type Network interface {
+	// a load balancing endpoint that libnetwork will not automatically
+	// remove when the remaining endpoints leave the network.  So, to remove
+	// an ingress network requires a special function declaring said intent.
+	DeleteIngress() error
```
Yeah, this is a choice of ugliness.  We can either change the libnetwork API in a backwards incompatible way by changing the Delete() signature or we can add a new method.   I was opting for the latter.  Will change the signature to make it more semantically clear.

ctelfer commented on network.go:
```diff
@@ -957,10 +975,36 @@ func (n *network) delete(force bool) error {
+		// If we got to this point, then the following must hold:
+		//  * force is true, OR
+		//  * endpoint count == 1 AND:
+		//      * n.ingress is false OR
```
Agree that the comment is confusing and superfluous.  Will simplify.  Agree that we don't need to mention ingress endpoint deletion:  it's explained above.

fcrisciani commented on service_linux.go:
```diff
@@ -102,13 +102,11 @@ func (n *network) addLBBackend(ip net.IP, lb *loadBalancer) {
-		return
-	}
-	if n.ingress && !sb.ingress {
+		logrus.Errorf("addLBBackend %s/%s: unable to find LB OS sandbox", n.ID(), n.Name())
```
In the past I remember these 2 to be very frequent, is it still the case? if this is not a fatal condition I would not use Error.

ctelfer commented:
Strangely, github refuses to show this comment in my web browser.  FWIW, I
have specifically looked for these returns during automated testing and
haven't seen it trigger yet.  A lack of load balancer sandbox will
definitely be an error when adding a binding because we are now locking the
network ID (see 2 commits prior) and testing for the presence of the
network before attempting to add load balancers.  The only way that sandbox
should be gone is if the network is being deleted.  But that also takes the
same network lock, so the two should never happen together.

As far as the osSbox condition, I don't mind downgrading that ...  It
should probably be at least warning level since it means that the load
balancing doesn't get programmed and this could mean complete service
disruption for single-container services.

On Fri, Jun 22, 2018 at 1:14 AM, Flavio Crisciani <notifications@github.com>
wrote:

> *@fcrisciani* commented on this pull request.
> ------------------------------
>
> In service_linux.go
> <https://github.com/docker/libnetwork/pull/2138#discussion_r197342147>:
>
> >  		return
>  	}
>  	if sb.osSbox == nil {
> -		return
> -	}
> -	if n.ingress && !sb.ingress {
> +		logrus.Errorf("addLBBackend %s/%s: unable to find LB OS sandbox", n.ID(), n.Name())
>
> In the past I remember these 2 to be very frequent, is it still the case?
> if this is not a fatal condition I would not use Error.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/libnetwork/pull/2138#pullrequestreview-131068245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AK8iXqbYPteTJqib5rINsv4NBLCwTiomks5t_H0cgaJpZM4Tadmu>
> .
>

fcrisciani commented on network.go:
```diff
@@ -1057,16 +1059,15 @@ func (n *network) delete(force bool, rmLBEndpoint bool) error {
-	c.cleanupServiceDiscovery(n.ID())
+	// Capture remaining binding info to clear after deleting
+	// the network from the store.
+	bindings = c.gatherBindingInfo(n.ID())
```
fetching the info here, and unfortunately because the gatherBinding requires the network, does not seem to give any more guarantee on the cleanup. I wonder if at this point we should just leverage the cleaning done from the networkDB

ctelfer commented on network.go:
```diff
@@ -2090,20 +2160,23 @@ func (n *network) createLoadBalancerSandbox() error {
 	}
-	return sb.EnableService()
+
+	err = sb.EnableService()
```
Very deliberately no.   This was the error that the patch needed to fix.  If one just does `return sb.EnableService()` then if `EnableService` returns an error that value does not get propagated to the function's `err` variable.   This, in turn, means that the cleanup defer functions will not detect the error and will thus not clean up after the error.

abhi replied:
ok then you can do this :) 
```
func (n *network) createLoadBalancerSandbox() (err error) {
```
This would keep the code clean. And avoid things like >>>>>>here

ctelfer replied:
That gets rid of the first line (err declaration for the function), but that's about it.  I'm not sure how that avoids things like `ep, err := n.CreateEndpoint(...)`.   The `:=` is (presumably) in the code because we don't want to declare the type of `ep`.  It's not to declare `err`.   Of course that has its own dangers.  If that statement were inside a loop, if-statement, switch statement or function body it would shadow the outer declaration causing disaster as well.   But that's true for named return values in the same way.  So I'm not sure what you're getting at.  Are you saying it should be changed to:
```
   var ep Endpoint
   ep, err = n.CreateEndpoint(...)
   ...
```

As far as the named return values and use of bare returns -- *shrug* it could be argued that this is less clear:
```
    x, err := foo()
    if err {
        return     // Looks at first glance like we aren't returning the error
    } 
```
and some prominent golang folks have argued that these should be used sparingly.  But I can put it in if desired.   The real issue is that the shadowing rules in golang are a bit of a misfeature.   One can see what they were aiming for, but there really aren't any clean-looking-while-still-robust patterns w/ regard to error handling because of them.  (although I'd be happy to learn otherwise)

abhi replied:
you can always do ```return err``` with the suggested syntax. It maps it to the defer value check and thats used in a lot of big golang code bases.

abhi replied:
```
func (n *network) createLoadBalancerSandbox() (retErr error) {
	sandboxName := n.lbSandboxName()
	sbOptions := []SandboxOption{}
	if n.ingress {
		sbOptions = append(sbOptions, OptionIngress())
	}
	sb, err := n.ctrlr.NewSandbox(sandboxName, sbOptions...)
	if err != nil {
		return err
	}
	defer func() {
		if retErr != nil {
			if e := n.ctrlr.SandboxDestroy(sandboxName); e != nil {
				logrus.Warnf("could not delete sandbox %s on failure on failure (%v): %v", sandboxName, err, e)
			}
		}
	}()

	endpointName := n.lbEndpointName()
	epOptions := []EndpointOption{
		CreateOptionIpam(n.loadBalancerIP, nil, nil, nil),
		CreateOptionLoadBalancer(),
	}
	if n.hasLoadBalancerEndpoint() && !n.ingress {
		// Mark LB endpoints as anonymous so they don't show up in DNS
		epOptions = append(epOptions, CreateOptionAnonymous())
	}
	ep, err := n.createEndpoint(endpointName, epOptions...)
	if err != nil {
		return err
	}
	defer func() {
		if retErr != nil {
			if e := ep.Delete(true); e != nil {
				logrus.Warnf("could not delete endpoint %s on failure on failure (%v): %v", endpointName, err, e)
			}
		}
	}()

	err := ep.Join(sb, nil)
	if err != nil {
		return err
	}

	return sb.EnableService()
}
```

ctelfer replied:
Ah, so a non-blank return statement assigns to the named return variable!  Perfect.  Much cleaner.  Will change.  Thanks!

abhi commented on network.go:
```diff
@@ -2073,11 +2132,15 @@ func (n *network) createLoadBalancerSandbox() error {
+		// Mark LB endpoints as anonymous so they don't show up in DNS
+		epOptions = append(epOptions, CreateOptionAnonymous())
+	}
 	ep, err := n.createEndpoint(endpointName, epOptions...)
```
>>>>>>here

fcrisciani commented:
@ctelfer can you also rebase?

ctelfer commented:
3 changes to the latest push:
  * revised createLoadBalancerSandbox()'s error handling per @abhi 's suggestion
  * also removed the RemoveLBEndpointAndDelete() in favor of passing an option (which can otherwise be omitted) instead.  This leaves the API backward compatible while avoiding adding a new method to the interface.
  * rebased to head of master

ctelfer commented:
For those interested, the moby patch to make use of this code (outside of vendoring) is quite thin:
```
diff --git a/daemon/network.go b/daemon/network.go
index 4263409..56d1067 100644
--- a/daemon/network.go
+++ b/daemon/network.go
@@ -4,7 +4,6 @@ import (
        "context"
        "fmt"
        "net"
-       "runtime"
        "sort"
        "strconv"
        "strings"
@@ -230,9 +229,7 @@ func (daemon *Daemon) releaseIngress(id string) {
                return
        }
 
-       daemon.deleteLoadBalancerSandbox(n)
-
-       if err := n.Delete(); err != nil {
+       if err := n.Delete(libnetwork.NetworkDeleteOptionRemoveLB); err != nil {
                logrus.Errorf("Failed to delete ingress network %s: %v", n.ID(), err)
                return
        }
@@ -349,7 +346,7 @@ func (daemon *Daemon) createNetwork(create types.NetworkCreateRequest, id string
                nwOptions = append(nwOptions, libnetwork.NetworkOptionConfigFrom(create.ConfigFrom.Network))
        }
 
-       if agent && driver == "overlay" && (create.Ingress || runtime.GOOS == "windows") {
+       if agent && driver == "overlay" {
                nodeIP, exists := daemon.GetAttachmentStore().GetIPForNetwork(id)
                if !exists {
                        return nil, fmt.Errorf("Failed to find a load balancer IP to use for network: %v", id)
@@ -512,37 +509,6 @@ func (daemon *Daemon) DeleteNetwork(networkID string) error {
        return daemon.deleteNetwork(n, false)
 }
```
plus remove the now superfluous `daemon.deleteLoadBalancerSandbox()` function.

orenye commented:
@ctelfer, can you please elaborate a bit this sentence: 

> First, the original behavior included the ability to load balance to a service by sending to any task IP address for the service.
What exactly can't the user do now?

ctelfer commented:
Hey @orenye ...  Sorry for the confusion, but that statement is actually erroneous and was caused by me misremembering an ingress processing scenario when I wrote it.  Users should be able to do everything with the new load balancing scheme that they could with the previous one.

orenye commented:
Hi @ctelfer , great, thanks!
